### DISCUSSION TITLE
valtio-like render optimization for zustand

### DISCUSSION THREAD

**Discussion by my-lalex** [pmndrs/zustand#901] [Ideas]
Hey there,

I'm a huge fan of both zustand and valtio, I think the DX is awesome.
The feature I love the most in valtio is its render optimization via proxy, but the fact that the state is deeply proxied may be a problem for my use case as I often use plain object comparison (`Object.is` or `===`) in my applications...

Using a different selector in each component is kinda heavy to use, so I created a `useZustand` hook (a kind of "zustio" mix?), taking a vanilla zustand store as parameter.

I'm not sure this hook fully cover the common issues zustand fixed, (zombie child problem, react concurrency, context loss). I used to create React apps with class based components and just starting to use hooks, so I'm not confident enough with their edge cases to create a PR... but I think some users might find this approach interesting.

So, I removed the whole "slice" code from the zustand hook, but maybe I was a bit too heavy on the "delete line" command?

```typescript
import { useEffect, useLayoutEffect, useReducer, useRef } from "react";
import { State, StoreApi } from "zustand";

// For server-side rendering: https://github.com/pmndrs/zustand/pull/34
// Deno support: https://github.com/pmndrs/zustand/issues/347
const isSSR =
    typeof window === "undefined" ||
    !window.navigator ||
    /ServerSideRendering|^Deno\//.test(window.navigator.userAgent);

const STATE_GETTER_PROPERTY = "_$zustandGetState";

const useIsomorphicLayoutEffect = isSSR ? useEffect : useLayoutEffect;

type Frozen<T> = T extends (...args: any[]) => any
    ? T
    : T extends Promise<infer V>
    ? Frozen<V>
    : { readonly [K in keyof T]: Frozen<T[K]> };

export const useZustand = <T extends State>(api: StoreApi<T>) => {
    const [, forceUpdate] = useReducer((c) => c + 1, 0);

    const state = api.getState();
    const stateRef = useRef(state);

    const accessed: Set<keyof T> = new Set();

    const proxyState = (state: T) => {
        const proxied = new Proxy(state, {
            get: function (t, p) {
                accessed.add(p as keyof T);
                return t[p as keyof T];
            },
        });
        return proxied;
    };

    // Replace get state so properties accessed via getState()
    // are tracked... even in state actions
    if (!api.getState.hasOwnProperty(STATE_GETTER_PROPERTY)) {
        const getOriginalState = api.getState;
        const getStateProxy = () => proxyState(getOriginalState.apply(api));

        // Create hidden property to identify if getState() is already returning a proxy
        Object.defineProperty(getStateProxy, STATE_GETTER_PROPERTY, {
            configurable: false,
            enumerable: false,
            value: getOriginalState,
        });
        api.getState = getStateProxy;
    }

    const stateBeforeSubscriptionRef = useRef(state);
    useIsomorphicLayoutEffect(() => {
        const listener = () => {
            const nextState = api.getState();
            let isNewSlice = false;

            for (const k of accessed) {
                if (!Object.is(nextState[k], stateRef.current[k])) {
                    // At least one of used props has changed
                    isNewSlice = true;
                    break;
                }
            }

            if (isNewSlice) {
                stateRef.current = nextState;
                forceUpdate();
            }
        };
        const unsubscribe = api.subscribe(listener);
        if (api.getState() !== stateBeforeSubscriptionRef.current) {
            listener(); // state has changed before subscription
        }
        return unsubscribe;
    }, []);

    return api.getState() as Frozen<T>;
};
```

Usage is

```typescript
import create from "zustand/vanilla";

const zu = create(() => ({
    prop1: 2,
    prop2: "hey !",
    updateProp1: (v: number) => {
        zu.setState({ prop1: v });
    },
    updateProp2: (v: string) => {
        zu.setState({ prop2: v });
    },
    getProp2: () => zu.getState().prop2,
}));

export const TestComponent = () => {
    const zus = useZustand(zu);

    return (
        <div>
            <span>Prop2: {zus.getProp2()}</span>
            <button
                onClick={() => {
                    // Won't trigger re-render as 'prop1' is not used
                    zus.updateProp1(10);
                }}
            >
                Change prop1
            </button>
            <button
                onClick={() => {
                    // Will trigger re-render as getProp2() is using getState().prop2
                    zus.updateProp2(zus.getProp2() + "!");
                }}
            >
                Change prop2
            </button>
        </div>
    );
};

export const TestComponentWithSpread = () => {
    const { prop1, updateProp1, updateProp2 } = useZustand(zu);

    return (
        <div>
            <span>{prop1}</span>
            <button
                onClick={() => {
                    // Will trigger re-render as 'prop1' used by
                    // { prop1, updateProp1, updateProp2 } = useExtendedZustand(zu);
                    updateProp1(10);
                }}
            >
                PROP1
            </button>
            <button
                onClick={() => {
                    // Will not trigger re-render as prop2 isn't used
                    updateProp2("bye!");
                }}
            >
                PROP2
            </button>
        </div>
    );
};
```

I would be glad to see this in zustand, and happy to help implementing it with an experienced hook coder. And also, if I could have feedback on what could be wrong with my implementation, if I'm missing something... or maybe tell me why this approach is not a good idea?

Big up to the `pmndrs` collective

**Comment by dai-shi** [comment]
Hi, nice try!

This have been requested once, and there's a support with [react-tracked](https://github.com/dai-shi/react-tracked).
See this example: https://react-tracked.js.org/docs/tutorial-zustand-01/
It supports nested tracking and tunes performance.

We want to keep zustand small (yes **really** small), so there's no plan to add such a hook. But, that doesn't stop someone to create a third-party library. I like the way you used `zustand/vanilla`. You might want to try using `use-sync-external-store`.

While this is out of the scope of zustand, let's leave this issue open for a while for further discussions.

**Comment by my-lalex** [comment]
Hi Daishi,

Thanks for your answer, and for leaving the issue open, I hope this could lead to interesting conversations.

I didn't go very deep in the `react-tracked` documentation, but it seems to me that it has the same "issue" than `valtio` (I know you wrote both): the deep proxying of data structures. So objects nested in state are proxies too, and comparing to objects who are supposed to be the same won't succeed as the proxy is created twice.... this won't be a problem in most use cases, but it is in mine. Maybe this could be fixed by a WeakMap (not creating a proxy twice for the same instance)...

Back to my implementation, there's a huge flaw in it: I create only one proxy, that will track accessed properties to only the first component. Those properties will be used to re-render every component using the api...  I'm currently investigating on how to fix this...

**Comment by dai-shi** [comment]
Deep proxying should work fine, but I'm not sure if I understand your issue. Can you try to create a small reproduction with zustand + react-tracked?

**Comment by my-lalex** [comment]
I made a quick mockup (please, don't judge, this has been made in 30mn)....

This a list of articles, available in different shops. The shops can be checked to enable a filter on this list.
I use a getShop factory to be sure I don't create twice the same instance of a Store. Usually I encapsulate this behaviour in an OOP class set.

The checkboxes don't get checked when I click on one because the test `shopFilter.includes(sh)` return false and it's the "expected" behaviour: `sh` is the original object, and `shopFilter` is a proxied array containing Proxies...

I know that there is alternatives: using the _id to compare or create a ".equals" method in the Shop class/object... but I think that being able to compare two objects from their reference is a good thing, and prevent memory leaks too...

If I switch from useTracked to useStore, everything works as expected...

```typescript
import { useEffect, useState } from "react";
import { createTrackedSelector } from "react-tracked";
import create from "zustand";


type Shop = {
    _id: string;
    name: string;
};

const storeRegistry: Record<Shop["_id"], Shop> = {};

const getShop = (_id: Shop["_id"], name?: string) => {
    return (storeRegistry[_id] ??= { _id, name: name || "Empty shop" });
};

const getAllShops = () => {
    return Object.values(storeRegistry);
};

type Article = {
    name: string;
    shops: Shop[];
};

type Store = {
    articles: Article[];
    shopFilter: Shop[];
    setShopFilter: (shops: Shop[]) => void;
    enableShopFilter: (shop: Shop, active: boolean) => void;
};

const useStore = create<Store>((set, get) => ({
    articles: [
        { name: "Bag", shops: [getShop("london", "London"), getShop("paris", "Paris"), getShop("tokyo", "Tokyo")] },
        { name: "Frame", shops: [getShop("london", "London")] },
        { name: "Umbrella", shops: [getShop("paris", "Paris"), getShop("tokyo", "Tokyo")] },
    ],
    shopFilter: [],

    setShopFilter: (shops: Shop[]) => {
        set({ shopFilter: [...shops] });
    },

    enableShopFilter(shop: Shop, active: boolean) {
        set((currentState) => {
            const currentShopFilter = currentState.shopFilter;
            if (currentShopFilter.includes(shop)) {
                if (!active) {
                    return { shopFilter: currentShopFilter.filter((s) => s != shop) };
                }
            } else if (active) {
                return { shopFilter: currentShopFilter.concat(shop) };
            }
            return currentState;
        });
    },
}));

const useTracked = createTrackedSelector(useStore);

export const TestComponent = () => {
    const { articles, shopFilter, enableShopFilter } = useTracked();
    const [filtered, setFiltered] = useState<Article[]>([...articles]);

    useEffect(() => {
        setFiltered(
            shopFilter.length
                ? articles.filter((a) => {
                      for (const articleShop of a.shops) {
                          if (shopFilter.includes(articleShop)) return true;
                      }
                      return false;
                  })
                : [...articles]
        );
    }, [articles, shopFilter]);
    return (
        <div>
            <div className="filter">
                {getAllShops().map((sh) => {
                    return (
                        <>
                            <input
                                type="checkbox"
                                checked={shopFilter.includes(sh)}
                                onChange={(e) => enableShopFilter(sh, e.currentTarget.checked)}
                            />
                            <span>{sh.name}</span>
                        </>
                    );
                })}
            </div>
            <ul className="articles">
                {filtered.map((a) => (
                    <li>
                        {a.name} ({a.shops.map((s) => s.name).join(",")})
                    </li>
                ))}
            </ul>
        </div>
    );
};
```

Actually, what I try to do is only to check if top level state properties should trigger a re-render. I'm ready to use Immer for nested properties (that's the price to pay if you want to use references). I just think the selector notation with the optional 'shallow' parameter is very heavy to use and felt in love with the valtio/react-tracked way...

**Comment by dai-shi** [comment]
It is true that there's a pitfall with memo/useMemo/deps with react-tracked/valtio, because property access is not stable across renders. So, it makes sense that, in your case, you want to proxy only at the top level.

That said, regardless of react-tracked/valtio, or even zustand, I think using `useEffect` to derive state is a bad practice in React in general. So, your specific case could be done without `useEffect`.

```js
    const { articles, shopFilter, enableShopFilter } = useTracked();
    const filtered = 
            shopFilter.length
                ? articles.filter((a) => {
                      for (const articleShop of a.shops) {
                          if (shopFilter.includes(articleShop)) return true;
                      }
                      return false;
                  })
                : [...articles];
```

This may not work because of `shopFilter.includes(sh)` as you said. Can you try and confirm?
(In such case, I would compute it outside React.)

**Comment by my-lalex** [comment]
Hi Daishi.

As you predicted, it still doesn't work because of the non-stable aspect of the data reference to a proxy...

I'm not sure I entirely understand why `useEffect` + `useState` is a bad practice, would you mind to give me a deeper explanation?
One thing I'm sure is that I cannot use your way because the filter can become kinda heavy (in real life, I work with a 250K articles array, each containing much more date than juste name on which many filters can be applied) and some other state change can cause a re-render, but maybe `useMemo` would be a better option?

On the re-render optimisation I tried to implement, I think my way of using proxies with zustand is a dead-end, as proxying getState can't be done for each component independently, and if the state is accessed internally by the store, the property access won't be tracked... maybe a good lead would be to add a syntactic sugar layer to valtio and which would use `ref()` each time I change a top-level field? I didn't push the investigation further (for now), but I think useSnapshot is creating another level of proxies just to track usage, I'm I right? I wouldn' know where to start...

It's another topic entirely and I don't know where I could talk about it with you and other users, but the way I'm using now with zustand is pretty "ugly". I think the selector/equalityFn can handle a lot of situations, but for simple use it leads for me as this kind of code (I'm really concerned by code readability and DX)
```typescript
store.subscribe(
    ({ articles }) => ({ articles }),
    ({ articles }) => {
        // Do my stuff
    },
    { equalityFn: shallow }
);
```

I'm trying to have a code more like the `useEffect` or `useState` API
```typescript
store.subscribe(
    ({ articles }) => {
        // Do my stuff
    },
    "articles" // or ["articles"]
);
```

For now, I created a zustand inspired little state management library to handle simple states, including a subscribeWithProps feature by default, and a hook taking optional "watched" props in parameter (just like the last pseudo-code). I return the full state, but the typescript's type is omitting unused properties to avoid using state properties which won't cause a re-render (and the full state type if no properties are given).... it's a bit twisted but that's the compromise I was ready to do....

```typescript
/** vanilla **/
const addListener = (
    listener: StateListener<StateType>,
    props?: (keyof StateType)[]
  ) => {...}

/** react hook **/
// Regular zustand-like store
export function useState<StateType extends State>(
  store: Store<StateType>
): StateType;
// With props specific render optimization
export function useState<
  StateType extends State,
  StateKey extends keyof StateType
>(store: Store<StateType>, ...props: StateKey[]): Pick<StateType, StateKey>;
```

**Comment by dai-shi** [comment]
> It's another topic entirely and I don't know where I could talk about it with you and other users

Yeah, there are no space to talk about across libraries. At least, you may want to start a new discussion to split from this one.

> On the re-render optimisation I tried to implement, I think my way of using proxies with zustand is a dead-end

It sounds like your understanding of the limitation of proxies is pretty good, and feels like bare zustand without proxies gives you more flexibility.

> I'm not sure I entirely understand why useEffect + useState is a bad practice, would you mind to give me a deeper explanation?

useEffect triggers another render, so it's technically slow.

```js
const [state, setState] = useState(...);
const [derivedState, setDerivedState] = useState(...);
useEffect(() => {
  setDerivedState(someFn(state)); // this will trigger another render
}, [state]);
// here, state and derivedState is not in sync
```

> but maybe useMemo would be a better option?

Yes, for your use case, it should be the best option. (alternative option is memoization outside react, `reselect` or valtio's `derive` or anything.)
